Toolchains: parse-time aliases for dependencies

becker33 opened the pull request:
This PR includes 2 primary components:

## Conditional dependencies

Conditional dependencies can be expressed with a `when=` clause on the dependency edge attributes

For example:

```
hdf5 %[when=+fortran virtuals=fortran] gcc ^[when=^mpi, virtuals=mpi ] mpich
```

means that hdf5 depends on mpich for mpi when it depends on mpi, and depends on gcc for fortran when the fortran variant is turned on.

You may need to quote this on the CLI, like so:

```console
> spack spec "hdf5 %[when=+fortran virtuals=fortran] gcc ^[when=^mpi, virtuals=mpi ] mpich"
```

## Toolchains

Toolchains are parse-time aliases for dependencies. This allows complex dependency relationships to be encoded by simple aliases. The toolchains feature is designed to facilitate the use-case previously allowed by configuring compilers to have the fortran compiler from a different compiler set -- toolchains are however more general, and can apply to any dependencies.

Toolchains are configured in the `toolchains` config section, e.g.

```yaml
toolchains:
  clang_gfortran: 
      - spec: %[virtuals=c] clang
        when: %c
      - spec: %[virtuals=cxx] clang
        when: %cxx
      - spec: %[virtuals=fortran] gcc
        when: %fortran
  gcc_mpich:
      - spec: %[virtuals=c] gcc
        when: %c
      - spec: %[virtuals=cxx] gcc
        when: %cxx
      - spec: %[virtuals=fortran] gcc
        when: %fortran
      - spec: %[virtuals=mpi] mpich
        when: %mpi
```

With this configuration, you can now use `%clang_gfortran` and `%gcc_mpich` as you would a compiler.  You can add flags, e.g.:

```yaml
gcc_mpich:
    - spec: cflags="-O3 -g"
    - spec: cxxflags="-O3 -g"
    - spec: fflags="-O3 -g"
    - spec: %[virtuals=c] gcc
      when: %c
    - spec: %[virtuals=cxx] gcc
      when: %cxx
    - spec: %[virtuals=fortran] gcc
      when: %fortran
    - spec: %[virtuals=mpi] mpich
      when: %mpi
```

The `clang_gfortran` toolchain is roughly equivalent to configuring a Spack v0.* compiler to have clang executables for c/cxx and gfortran for fortran. The `gcc_mpich` toolchain uses only gcc, but is extended to also include the mpi implementation to use.

You can currently only use `%` (direct dependency) in toolchain conditions, though we are working on enabling `^` conditions in a later PR.

You *can* write the same toolchains with conditional dependencies, but it's much harder to read:

```yaml
toolchains:
  clang_gfortran: "%[when='%c' virtuals=c]clang %[when='%cxx' virtuals=cxx]clang %[when='%fortran' virtuals=fortran]gcc"
  gcc_mpich: "%[when='%c' virtuals=c]gcc %[when='%cxx' virtuals=cxx]gcc %[when='%fortran' virtuals=fortran]gcc %[when='%mpi' virtuals=mpi] mpich"
```

## Some notes on  using toolchains:
- You likely want to use conditional dependencies as shown above. If you use unconditional dependencies, it will turn on variants to ensure the compiler can be used, and will fail for packages that don't depend on a given language.
- For similar reasons, it's best to separately specify each condition for shared virtuals like c/cxx/fortran from the same package.

TODO:
- [x] unit tests for prefers/requires with conditional dependencies
- [x] unit tests for constraints with conditional dependencies (partially tested by concretization tests)
- [x] unit tests for satisfies with conditional dependencies (partially tested by concretization tests)
- [x] documentation

becker33 committed:
initial wip with concretization for conditional dependencies

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spec.long_spec: update for conditional deps

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
SpecParsingError: only print context if available

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
conditional deps: bugfix for type mismatch in _add_dependency

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup 2355eb5c3e

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
satisfies for conditional dependencies

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spack toolchains

Add parser and config support for toolchains. A toolchain is a configured
string alias applied at parse time. When parsing the '%' sigil the parser
will match the token after the sigil to any toolchain name, and if there is
a match it will replace the sigil and following token with the tokens for
the associated string.

For example, the following config allows for the user to enforce usage of
clang for c/cxx and gcc for fortran with a single option `%my_toolchain`

toolchains:
  my_toolchain: ^[when=%c virtuals=c ]clang ^[when=%cxx virtuals=cxx ]clang ^[when=%fortran virtuals=fortran ]gcc

Note that the space before the ']' symbol is required for proper parsing. This
is considered a bug and will be fixed when possible.

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
toolchains/conditional deps: support prefers/requires

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
constrain: handle conditional deps

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
conditional deps: add parsing test

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spec: filter conditional edges in Spec.dependencies/edges_to_dependencies

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spec.edge_attributes: deduplicate dependency conditions from multiple parents

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spec.constrain: separate constraints from edges with different conditions

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spec.satisfies: Remove optimization that fails for spec with no dependency satisfying failed conditional dep

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
add test for toolchains

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
rebase fixup

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup rebase missing closing paren

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
parseable_tokens: fix mypy annotations

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
conditional deps: fix parsing tests

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
mypy fix

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
update completions

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
conditional deps: tests for Spec.satisfies and Spec.constrain

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
tests for conditional deps in requirements

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
add test file for conditional dependencies

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
add missing test_repo package

Signed-off-by: Gregory Becker <becker33@llnl.gov>

alalazo commented on lib/spack/spack/schema/toolchains.py:
```diff
@@ -0,0 +1,23 @@
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""Schema for repos.yaml configuration file.
```
There is a typo in the docstring (`repo.yaml` instead of `toolchains.yaml`)

alalazo commented on lib/spack/spack/schema/toolchains.py:
```diff
@@ -0,0 +1,23 @@
+"""Schema for repos.yaml configuration file.
+
+.. literalinclude:: _spack_root/lib/spack/spack/schema/toolchains.py
+   :lines: 14-
```
(minor) The line is likely wrong too...

alalazo commented on lib/spack/spack/schema/toolchains.py:
```diff
@@ -0,0 +1,23 @@
+from typing import Any, Dict
+
+#: Properties for inclusion in other schemas
+properties: Dict[str, Any] = {"toolchains": {"type": "object", "default": {}}}
```
We can use `default:` to check the type of the entry. By reading the PR description I assume toolchains are always a `Dict[str, str]`?

becker33 replied:
Currently yes, but with Todd's suggestions below there will also be an option for them to be

```
Dict[str, List[Dict[str, str]]
```

alalazo replied:
Apologies, I meant we can use `additionalProperties` (not `default:`!) to better specify the schema. No need for pattern properties and regex matching.

alalazo replied:
Done in ee48052fe825a97517fc86a90d81c768cc3978d0

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3200,6 +3252,9 @@ def setup(
         self.gen.h1("Spec Constraints")
         self.literal_specs(specs)
 
+        self.trigger_rules()
```
Why was this addition necessary here?

becker33 replied:
Because `self.literal_specs()` now creates conditions for conditional dependencies, and those conditions need to be rendered out

tgamblin replied:
add a comment saying that this is for `literal_specs()`

alalazo replied:
This is also wrong. We should modify `self.literal_specs()` instead

alalazo replied:
Fixed in https://github.com/spack/spack/pull/50481/commits/11e71c5f63713b2e83a68f0a6c93ae9e0615ae83

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1877,7 +1882,7 @@ def _get_condition_id(
         return cond_id
 
-    def condition(
+    def condition_clauses(
```
Splitting `condition` into `condition_clauses` + `condition` seems unrelated, and can be extracted into its own PR (I see no use of `condition_clauses` other than in `condition`).

We also miss the type-hint for the return value.

alalazo replied:
Done in #50597

becker33 replied:
Yeah I can undo this, that was vestigial from an early attempt at implementing the conditional deps.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3333,11 +3388,49 @@ def literal_specs(self, specs):
 
         self.effect_rules()
 
+    def generate_conditional_dep_conditions(self, spec, condition_id):
```
If I read this correctly, the function seems to be O(edges**2). Wondering if we can improve on that. Also, would be good to have a docstring, and type hints.

becker33 replied:
I've put some thought into it and I'm not seeing a better algorithm, but if you have an idea I'm definitely open to it.

Technically this is O(edges * conditional_edges), which should be substantially smaller than edges**2. Most importantly that restricts it to O(edges) for concrete specs, and abstract specs are relatively small and less dense in edges.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -429,8 +429,23 @@ trigger_and_effect(Package, TriggerID, EffectID)
 % is derived. This allows imposed constraints to be canceled in special cases.
+
+% Effects of direct conditions hold if the trigger holds
+impose(EffectID, node(X, Package))
```
(minor) This is probably worth simplifying to avoid duplicating the body of the rule. The purpose seems similar to `activate_requirement`, so should we have an `activate_effect`?

alalazo replied:
Already fixed.

alalazo commented on lib/spack/spack/test/concretization/conditional_dependencies.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@pytest.mark.parametrize("holds,mpi", [(True, "zmpi"), (True, "mpich"), (False, "mpich")])
+def test_conditional_deps(holds, mpi, config, mock_packages):
```
Can you add a brief docstring explaining the use cases we are testing?

alalazo commented on lib/spack/spack/test/concretization/errors.py:
```diff
@@ -81,8 +81,10 @@ def test_internal_error_handling_formatting(tmp_path):
-        "    - foo+x\n"
-        "    - bar+y\n"
+        "    - input: foo+x\n"
+        "      output: foo@=1.0~x\n"
```
This change seems just related to using `long_spec` in the error message. We can have a separate PR for that, of just a few lines.

alalazo replied:
Done in #50597

alalazo commented on lib/spack/spack/test/concretization/requirements.py:
```diff
@@ -1301,3 +1302,52 @@ def test_requirements_on_compilers_and_reuse(
+    )
+    abstract = spack.spec.Spec(abstract)
+
+    # Configure two gcc compilers that could be concretized to
```
It seems a lot of the setup is shared with tests in `conditional_dependencies.py`, and we can factor that together.

alalazo commented on lib/spack/spack/test/conftest.py:
```diff
@@ -2054,7 +2054,7 @@ def shell_as(shell):
 def nullify_globals(request, monkeypatch):
     ensure_configuration_fixture_run_before(request)
-    monkeypatch.setattr(spack.config, "CONFIG", None)
+    monkeypatch.setattr(spack.config, "CONFIG", {})  # So basic get operations do not throw
```
This is wrong, the fixture is there to ensure we do not rely on globals in some tests. If the change is necessary, it's a regression (something that before didn't rely on CONFIG, now relies on it).

becker33 replied:
Well yeah, everything that parses a spec now relies on config -- but by leaving it as an empty dict, we still ensure that it doesn't rely on any particular value.

alalazo replied:
We rely on having a global empty dict, and failing without... which seems weird.

> Well yeah, everything that parses a spec now relies on config

From a design perspective, I think this is bad: now to parse anything, say `hdf5`, we need to have a bunch of yaml files available, and loaded in memory.

Given that toolchains are always introduced by `%`, and are thus a leaf node on an abstract spec DAG[^1], we should be able to leave the parser untouched, and expand the toolchain leaf as a second step.

We might not do it now, due to lack of time, but the change above signals a new tech debt imo. 

[^1]: At least until we implement another pair of symbols to have recursive definitions of specs

tgamblin replied:
Agree with @alalazo that toolchains should be abstractified, but not right now -- I don't think we have time.

alalazo replied:
So, I addede thi https://github.com/spack/spack/pull/50481/commits/d0af98878e3cd96f6fd898611a4be0241f3dc5b1 which is a minimal change to avoid relying on having `CONFIG` available. I also added a FIXME.

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -573,6 +593,14 @@ def test_abstract_specs_with_propagation(self, lhs, rhs, expected, constrained):
         c.constrain(lhs)
         assert c == constrained
 
+    def test_satisfies_conditional_dep(self):
```
(minor) We should probably document better which cases are tested here.

alalazo replied:
https://github.com/spack/spack/pull/50481/commits/cf7c7b351925181e0b2bd77b67200229ec55b6bb

alalazo commented on lib/spack/spack/main.py:
```diff
@@ -1090,12 +1090,13 @@ def _handle_solver_bug(
-            + "\n    - ".join(str(s) for s, _ in wrong_output)
-            + "\n    Please report a bug at https://github.com/spack/spack/issues"
-        )
+        msg = "internal solver error: the following specs were concretized, but do not satisfy "
```
Changes to this file can also be extracted in a separate PR dealing with error output.

alalazo replied:
Done in #50597

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -2085,14 +2138,14 @@ def long_spec(self):
             # note: depflag not allowed, currently, on "direct" edges
             edge_attributes = ""
-            if item.virtuals:
+            if item.virtuals or item.when != Spec():
```
(minor) Wondering if we should have a sentinel object, instead of `Spec()`, e.g.:
```python
if item.when != NO_SPEC_CONDITION:
```
or similar.

tgamblin replied:
Sentinel or `None` is better IMO -- see above.

alalazo replied:
Leaving this one for a cleanup, in a following PR

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -3411,6 +3441,11 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
         mock_nodes_from_old_specfiles = set()
         for rhs_edge in other.traverse_edges(root=False, cover="edges"):
+            # Skip checking any conditional edge that is not satisfied
+            if rhs_edge.when != Spec() and not self.satisfies(rhs_edge.when):
```
By just reading the code, it seems to me the very fundamental assumption on subset semantic:
```
abstract_spec.satisfies(x) is True -> concrete_spec.satisfies(x) is True 
```
doesn't hold anymore for conditional deps. A very simple reproducer:
```python
Spack version 1.0.0.dev0
Python 3.13.0, Linux x86_64
>>> from spack.spec import Spec
>>> s = Spec("hdf5")
>>> constrain = Spec("%[when='+mpi' virtuals=zlib-api] zlib")
>>> s.satisfies(constrain)
True
>>> s.concretize()
==> Warning: <console>:1: `Spec.concretize` is deprecated and will be removed in version 1.0.0. Use `spack.concretize.concretize_one` instead.
>>> s.satisfies(constrain)
False
```
Did we work out the consequences of this ?

tgamblin replied:
We agreed, for the first draft of this, that whether or not conditional deps hold in `satisfies()` is less important than getting a working version of toolchains. I think we need to work this one out later -- post 1.0.

haampie replied:
Just to add to this, conditional specs act as a filter on the set of solutions just like abstract specs do. As an example, suppose the set of all packages consists of `foo` with a boolean variant `x` and `bar` without any variants.

The abstract spec `foo` induces the set of concrete specs:

```
{foo ~x %bar, foo +x %bar, foo ~x, foo +x}
```

The conditional spec

```
foo %[when=+x] bar
```

induces the following set of concrete specs

```
{foo ~x %bar, foo +x %bar, foo ~x}
```

where `foo +x` is absent due to the conditional constraint.

So, `foo` does **not** satisfy `foo %[when=+x] bar`. The right-hand side is more constrained than the left-hand side.

Finally, there is the case where the right-hand side is *not* constraining. For example `satisfies(foo ~x ~y, foo %[when='~x +y'] bar)` is vacuously true, since the when condition does not hold.

So, the logic for `satisfies(<lhs> spec, <when> spec => <consequence> spec)` should be:

```python
if intersects(lhs, <when>) and not satisfies(lhs, <consequence>):
    return False
```

The `intersects` check deals with the "vacuous truth" case, the `satisfies` check deals with the solution set constraint.

It's important to get this right.

alalazo replied:
I guess, to summarize, we should split the case of:
```
<when condition applies>
```
into:
```
<when condition cannot ever apply> -> ok to drop edge
<when condition may not apply> -> return False
```
?
Basically, differentiate the two abstract specs:
```
hdf5
hdf5~mpi
```
when tested against:
```
%[when=+mpi] <condition>
```

alalazo replied:
To differentiate the two cases we probably need:
```
if not spec.intersects(when_condition):
   # fine to skip edge, it's the hdf5~mpi case
elif not spec.satisfies(<condition>):
   # This should be False
   return False
```

alalazo replied:
Should be fixed in https://github.com/spack/spack/pull/50481/commits/0db9ec21aac540c26a4519eb3dd5869144ba1549

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -3468,6 +3503,7 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
                 for lhs_edge in self.traverse_edges(
                     root=False, cover="edges", deptype=("link", "run")
                 ):
+                    # TODO: do we need to avoid conditional edges here
```
See comment above

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -3484,18 +3520,30 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
+        lhs_nodes = list(self.traverse(root=False)) + sorted(mock_nodes_from_old_specfiles)
+        for rhs in other.traverse(root=False):
+            # Possible lhs nodes to match this rhs node
+            lhss = [lhs for lhs in lhs_nodes if lhs.satisfies(rhs, deps=False)]
```
`lhss`?

alalazo commented on lib/spack/spack/spec_parser.py:
```diff
@@ -189,6 +200,16 @@ def accept(self, kind: SpecTokens):
     def expect(self, *kinds: SpecTokens):
         return self.next_token and self.next_token.kind in kinds
 
+    def push(self, token_stream: Iterator[Token]):
```
Hmmm, this seems very coupled with the global CONFIG Wondering if we could keep the structure more modular.

alalazo commented on lib/spack/spack/spec_parser.py:
```diff
@@ -268,6 +291,9 @@ def add_dependency(dep, **edge_properties):
                 raise SpecParsingError(str(e), self.ctx.current_token, self.literal_str) from e
 
+        # Get toolchain information outside of loop
+        toolchains = spack.config.CONFIG.get("toolchains", {})
```
(minor) It would be good if we could avoid this here

tgamblin replied:
I think we want the toolchain parts of parser to be in a separate file (e.g. `toolchain.py`), as I suspect it'll grow.  e.g., we likely want something like this to clean up `spack spec` output (maybe not for 1.0 but eventually) and that can go in the dedicated toolchain module.

alalazo commented on lib/spack/spack/spec_parser.py:
```diff
@@ -528,6 +563,11 @@ def parse(self):
+
+        # Turn "when" into a spec
+        if "when" in attributes:
+            attributes["when"] = spack.spec.Spec(attributes["when"][0])
```
Isn't the Spec constructor just calling some function in the parser? Can we use `parse_one_or_raise`?

alalazo commented on lib/spack/spack/spec_parser.py:
```diff
@@ -573,8 +613,9 @@ class SpecParsingError(spack.error.SpecSyntaxError):
         message += f"\n{text}"
-        underline = f"\n{' '*token.start}{'^'*(token.end - token.start)}"
-        message += color.colorize(f"@*r{{{underline}}}")
+        if token:
```
In which case we don't have a token?

becker33 replied:
If a syntax error causes the EdgeAttributeParser to consume the portion of the string that should be tokenized as the dependency.

alalazo reviewed (changes requested):
I gave a first fast read to the PR. Will test drive it soon. A few comments:

1. There seems to be a violation of the subset semantic by conditional specs, see the comment below.
2. I think toolchains should be expanded only from `%`, and if possible the expansion shouldn't be so closely tied to the parser. Right now the spec parser depends on CONFIG, before it didn't.
3. The motivation to only expand from `%` is that we are otherwise making `^` position dependent in the spec, whereas before it was not (e.g. `foo ^bar ^toolchain` might be different from `foo ^toolchain ^bar`, which is highly unexpected)

I also wonder how the `when=` argument that we added in many Spec related methods might relate to conditional dependencies expressed with `depends_on(..., when=...)` in packages, but that might be postponed to a post v1.0 discussion, assuming we do not expose these new methods to the package API (need to double check that).

I pointed out a few opportunities to extract trivial refactors, and reduce the amount of changes in this PR "unrelated" to the feature.

tgamblin commented:
I agree with @alalazo's points, and I think we should restrict it just to `%` like he says.  A couple other high level comments:

You say:
> Toolchains are parse-time aliases for dependencies.
but I think users are going to want flags and maybe variants in here too.  Is that possible?

tgamblin commented:
I guess my other major comment is that the quoting here is going to put users off:

```
spack spec hdf5 %[when=\'+fortran\' virtuals=fortran] gcc ^[when=\'^mpi\', virtuals=mpi ] mpich
```

How can we avoid the \'?  I don't think that should be necessary.  If the user can quote the whole thing it's much better:

```
spack spec "hdf5 %[when='+fortran' virtuals=fortran] gcc ^[when='^mpi', virtuals=mpi ] mpich"
```

but I don't really get why +fortran or ^mpi should require quotes. Are you just quoting that *in case* there are spaces in the when spec?

becker33 committed:
schema/toolchains: fix up typos

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
parser: fixup ambiguous variable name

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
parser: use parse_one_or_raise instead of Spec()

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
parser: use Spec.constrain instead of pushing tokens back to stream

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
refactor tests for conditional dependencies

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
solver/asp::generate_conditional_dep_conditions add docstring/improve message

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
traverse.sort_edges: fix edge case for duplicate edges

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
re-add missing test package after rebase

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup conditional-languages test package location

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup conditional-languages test package import

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove vestigial import

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
regression test for non-dependency spec attributes in toolchains

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
update toolchains config format to allow complex config

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
rework fixture to avoid module namespace clash

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
move fixture to conftest.py

Signed-off-by: Gregory Becker <becker33@llnl.gov>

alalazo committed:
Update command completion

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
toolchains schema: use "additionalProperties"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
Trying to test drive the PR in different scenario. First failure:
```yaml
spack:
  specs:
  - hdf5 %clang_gfortran
  toolchains:
    clang_gfortran: "%[when='^c' virtuals=c] clang %[when='^cxx' virtuals=cxx] clang %[when='^fortran' virtuals=fortran] gcc"
  view: true
  concretizer:
    unify: true
```
This gives:
```console
$ spack concretize -f --fresh
==> Error: /tmp/spack-rk18t8z1/spack.yaml:2: Additional properties are not allowed ('toolchains' was unexpected)
```

alalazo committed:
Allow "toolchains" section in spack.yaml

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/schema/toolchains.py:
```diff
@@ -0,0 +1,40 @@
+        "default": {},
+        "additionalProperties": {
+            "oneOf": [
+                {"type": "string"},
```
I would remove this way of doing things from config, because it can lead to confusing specs, and possibly to subtle issues. For instance:
```yaml
toolchains:
  foo: "^[when=^mpi virtuals=mpi] mpich %[when=%c virtuals=c] gcc"
```
is different from:
```yaml
toolchains:
  foo: "%[when=%c virtuals=c] gcc ^[when=^mpi virtuals=mpi] mpich"
```
as in the first case the conditional compiler is parsed on `mpich`.

alalazo replied:
We'll allow this, but only document the other syntax.

alalazo reviewed (changes requested):
There are issues with white spaces in the parser:
![Screenshot from 2025-05-29 17-21-42](https://github.com/user-attachments/assets/c5e0917b-ec3c-4aa8-95fa-e1e34adb0b51)

becker33 committed:
allow % and ^ in unquoted key-value pairs

Signed-off-by: Gregory Becker <becker33@llnl.gov>

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -731,12 +731,14 @@ def __init__(
         self.depflag = depflag
         self.virtuals = tuple(sorted(set(virtuals)))
         self.direct = direct
+        self.when = when or Spec()
```
I'd rather set this initially to `None` to avoid extra object construction for every dependency on every spec -- does it hugely complicate things to do that?  I realize it's nicer programmatically to use a bare `Spec()`

becker33 replied:
It requires adding a conditional everywhere we check `edge.when.satisfies()` or call `edge.when.constrain()`. I don't think calling the spec constructor is relatively that expensive.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1050,6 +1058,9 @@ def select(
                 selected = (dep for dep in selected if any(v in dep.virtuals for v in virtuals))
 
+        if when is not None:
+            selected = (dep for dep in selected if dep.when == when)
```
Why do `dep.when == when` here and not `dep.when.satisfies(when)`?  I think the latter makes a lot more sense given the way we typically treat specs in the internal APIs.  I could also see someone asking for, say, all the conditional deps that would apply to versions `@4:` -- and you'll miss the ones that are just `@4` with `==`.

Only thing I could think of this affecting is if you call select or some other method that calls with with `when=Spec()` to get the unconditional deps, but it looks like you check that explicitly with `dep.when == Spec()` everywhere it would be needed in this PR.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1618,39 +1634,50 @@ def edge_attributes(self) -> str:
             name: filter dependencies by package name
             deptype: allowed dependency types
             virtuals: allowed virtuals
+            when: condition on conditional dependency or Spec() for unconditional
```
I find this description really hard to parse.  how about:

```suggestion
            when: optional spec to select matching dependency conditions
```

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1737,31 +1764,47 @@ def _set_architecture(self, **kwargs):
             depflag: dependency type for this edge
             virtuals: virtuals on this edge
             direct: if True denotes a direct dependency (associated with the % sigil)
+            when: if non-None, condition under which dependency holds
```
```suggestion
            when: optional condition under which dependency holds
```

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -430,8 +430,23 @@ trigger_and_effect(Package, TriggerID, EffectID)
+  :- pkg_fact(Package, condition_effect(SubconditionId, EffectID)),
+     subcondition(SubconditionID, ConditionID),
+     condition_holds(ConditionID, node(X, Package)),
+     trigger_and_effect(Package, TriggerID, EffectID),
```
See @alalazo's comment above -- isn't this a textbook case for a conditional rule body?

I think you could reduce this to a single rule (delete the first one) like so:

```suggestion
  :- pkg_fact(Package, condition_effect(SubconditionId, EffectID)),
     condition_holds(ConditionID, node(X, Package)) : subcondition(SubconditionID, ConditionID),
     trigger_and_effect(Package, TriggerID, EffectID),
```

tgamblin commented on lib/spack/spack/traverse.py:
```diff
@@ -35,8 +35,10 @@ class EdgeAndDepth(NamedTuple):
 
 
+# Sort edges by name first, then abstract hash
+# Sort by full edge string to break ties
```
```suggestion
# Sort edges by name first, then abstract hash, then full edge string to break ties
```

tgamblin commented on lib/spack/spack/test/spec_syntax.py:
```diff
@@ -882,6 +896,59 @@ def test_cli_spec_roundtrip(args, expected):
 
 
+@pytest.mark.parametrize(
+    "spec_str,toolchain,expected_roundtrip",
```
minor: I think we should start expanding these:

```suggestion
    ["spec_str", "toolchain", "expected_roundtrip"],
```

alalazo replied:
Let's leave this for later - no time atm for style issues like this one :slightly_smiling_face:

tgamblin commented on lib/spack/spack/test/env.py:
```diff
@@ -995,3 +995,67 @@ def test_using_multiple_compilers_on_a_node_is_discouraged(tmp_path, mutable_con
+
+
+@pytest.mark.parametrize(
+    "spack_yaml,expected,not_expected",
```
minor: expand
```suggestion
    ["spack_yaml", "expected", "not_expected"],
```

alalazo replied:
Why are we doing this? This is allowed, but actually no example in the entire pytest docs uses a list of strings. Wondering what we gain by doing things differently from what we do already.

becker33 committed:
add comment on why we call self.trigger_rules after literals

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fix syntax errors in tests

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
spec.py: tighten up language in docstrings for conditional deps

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
traverse.py: reword comment

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
update new tests to present parameter names as list

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
solver/concretize: rework imposition rule

Signed-off-by: Gregory Becker <becker33@llnl.gov>

haampie reviewed (dismissed):
blocking this over satisfies semantics https://github.com/spack/spack/pull/50481#discussion_r2120402910

alalazo committed:
Fix semantic of conditional dependencies

Now subset semantic is satisfied

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Push call to "trigger_rules" into "literal_specs"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix Spec.satisfies("%<virtual>")

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add more unit tests for conditional deps

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Minimal simplification in test setup

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Don't require CONFIG to parse a spec

This is a minimal change, that doesn't extract the
toolchain expansion into a separate operation.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix ^ -> % in tests, to reflect real use cases

Also add conditional branches on the same leaf,
under two different conditions

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Improve test satisfies conditional dep

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix satisfies semantic when % conditional edges are on the lhs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix satisfies semantic when ^ conditional edges are on the lhs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -2243,11 +2243,12 @@ def test_satisfies_and_subscript_with_compilers(default_mock_concretization):
 
     # Transitive build dependencies
-    assert s.satisfies("^gmake")
+    assert not s.satisfies("^gmake")
```
This was an actual bug in our `Spec.satisfies` on `develop`

becker33 committed:
Spec.edges_to_dependencies: revert addition of when argument

Signed-off-by: Gregory Becker <becker33@llnl.gov>

alalazo committed:
Don't mutate objects in memory that may be referred elsewhere

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Emit "direct_dependency" facts from conditional dependencies

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a unit test for a single matrix using a toolchain

Co-authored-by: Peter Scheibel <scheibel1@llnl.gov>
Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3134,20 +3134,6 @@ def setup(
         self.pkgs = node_counter.possible_dependencies()
         self.libcs = sorted(all_libcs())  # type: ignore[type-var]
 
-        # Fail if we already know an unreachable node is requested
-        for spec in specs:
-            # concrete roots don't need their dependencies verified
-            if spec.concrete:
-                continue
-
-            missing_deps = [
-                str(d)
-                for d in spec.traverse()
-                if d.name not in self.pkgs and not spack.repo.PATH.is_virtual(d.name)
-            ]
-            if missing_deps:
-                raise spack.spec.InvalidDependencyError(spec.name, missing_deps)
```
This was failing on:
```
foo %[when=%mpi virtuals=mpi] mpich
```
if `foo` couldn't depend on `mpi`. Since the check has been rarely useful, I removed it instead of making it more complicated by iterating on edges.

alalazo committed:
Add a unit test for a matrix using a different toolchains

Co-authored-by: Peter Scheibel <scheibel1@llnl.gov>
Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a unit test using a toolchain as a default requirement

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a unit test using a toolchain as a default preference

XFAIL for now

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a unit test using two toolchains in the same input spec

XFAIL for now

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

becker33 committed:
Revert "solver/concretize: rework imposition rule"

This reverts commit 4ce5188be450b23bdb2d1ce617ceda74493906ce.

becker33 committed:
test/env.py: fix satisfaction checks in xfail test to only test those we guarantee

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
spec_parser: remove vestigial docstring reference to pushing back tokens

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Spec.satisfies: ignore conditional edges in parent-aware manner

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
allow subconditions to be on separate nodes from primary condition

Signed-off-by: Gregory Becker <becker33@llnl.gov>

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -3438,6 +3438,12 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
             # The condition cannot be applied in any case, skip the edge
             if not self.intersects(rhs_edge.when):
                 continue
+            if (
+                rhs_edge.parent.name
+                and rhs_edge.parent.name in self
+                and not self[rhs_edge.parent.name].intersects(rhs_edge.when)
+            ):
+                continue
```
I'll look into this. Doesn't seem entirely correct at a first glance. For instance, we already checked with `self` above. Also, as a note for the future, it would have been good to have a new unit test that was failing before the change, and succeeding after the change to avoid regressions and see clearly which cases were buggy.

alalazo replied:
Indeed this implementation was buggy. Should be fixed in https://github.com/spack/spack/pull/50481/commits/c1acf1e2fbe3f0469024c1a6051c80f5f96c20cc that also adds a unit test

becker33 replied:
There was an xfail test that passed now that it was added, that's why I was working on it in the first place.

alalazo committed:
Add unit tests for multiple conditional specs on the same input string

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/spec_parser.py:
```diff
@@ -294,6 +310,13 @@ def add_dependency(dep, **edge_properties):
+                # Look ahead to match upcoming value to list of toolchains
+                if self.ctx.current_token.value == "%" and self.ctx.next_token.value in toolchains:
+                    assert self.ctx.accept(SpecTokens.UNQUALIFIED_PACKAGE_NAME)
+                    self._apply_toolchain(current_spec, toolchains[self.ctx.current_token.value])
```
We can reduce parsing a lot, if we store already parsed toolchains. Specs like:
```
foo %toolchain ^ bar %toolchain ^ baz %toolchain
``` 
are currently parsing `%toolchain` 3 times.

alalazo replied:
This requires a refactor of some kind though, because toolchains are currently in the parser, and that means we can't parse them as specs without triggering infinite recursion :worried:

becker33 replied:
This is fixed in 47030ae

alalazo committed:
Fix two bugs in conditional deps + unit tests

* Test the correct rhs edge in satisfies
* Do not fail concretization if a conditional dep
  refers to a non-existing variant

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Mark a new regression xfail

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Restrict toolchain definition to only allow % sigil

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix matrix expansion with when-possible unification

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

becker33 committed:
toolchains: avoid parsing the same toolchain repeatedly

Signed-off-by: Gregory Becker <becker33@llnl.gov>

alalazo committed:
Rework requirements to introduce an intermediate fact

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/spec_parser.py:
```diff
@@ -360,25 +361,34 @@ def _parse_node(self, root_spec):
                 edge.when.constrain(when)
 
-            spec.constrain(constraint)
+            self.parsed_toolchains[name] = toolchain
```
I think this is wrong _within_ the loop, the cached toolchain seems to be now only the last 
```yaml
- spec: ...
  when: ...
```
entry. I'll fix that in a following commit.

becker33 replied:
Oh ugh yeah that's definitely wrong. Let me know if you want me to fix it instead.

alalazo replied:
Parser fixed in a6665aaf280d795995df88212e4752989c410fc2

alalazo committed:
Fix parser, add unit test to avoid regression

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
To be done in following PRs (likely after v1.0):
- Move toolchain expansion outside of the parser: https://github.com/spack/spack/pull/50481#discussion_r2121709991
- Use a sentinel / function instead of `Spec()` to test for unconditional branches: https://github.com/spack/spack/pull/50481#discussion_r2099624806

becker33 committed:
Add docs for toolchains and conditional deps

Also cleaned up outdated references to compilers as separate from
dependencies and ordering of example specs

Signed-off-by: Gregory Becker <becker33@llnl.gov>

alalazo commented on lib/spack/docs/getting_started.rst:
```diff
@@ -689,6 +689,59 @@ Or it can be set permanently in your ``packages.yaml``:
+.. code-block:: yaml
+
+   toolchains:
+     gcc_all: cflags=-O3 '%[when=%c virtuals=c]gcc %[when=%cxx virtuals=cxx]gcc %[when=%fortran virtuals=fortran]gcc'
```
Should we document this one? I'd rather avoid that, cause when we'll extend toolchains to allow `^` it could lead to subtle issues due to:
```
%[...] gcc ^[...] mpich
^[...] mpich %[...] gcc
```
being different constraint. Using the other syntax, this cannot happen.

becker33 replied:
I still really prefer that syntax. It's more concise, represents each toolchain as a single line, and presents the toolchains the same way they will appear on abstract specs printed after parsing.

tgamblin replied:
Single line is *bad* for this.

alalazo commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1160,6 +1165,14 @@ configuration:
    mpileaks ^callpath@1.0 ^libelf@0.8.3
    mpileaks ^libelf@0.8.3 ^callpath@1.0
 
+Direct dependencies specified with ``%`` also differ from general
+dependencies because they associate with the most recent node, rather
+than with the root of the DAG. So in the spec ``root ^dep1 ^dep2
+^dep3`` all three dependencies are associated with the package
+``root``, but in the spec ``root ^dep1 %dep2 %dep3`` the spec
+``%dep2`` is associated with ``dep1`` and the spec ``%dep3`` is
+associated with ``dep2``.
```
Just a note that this is wrong. I'm fixing this in #50932
